### PR TITLE
ProcMount 1.30 update

### PR DESCRIPTION
**Pull request by haircommander**
KEP: https://github.com/kubernetes/enhancements/pull/4265

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
Hey @haircommander 👋 please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday March 12th 2024 18:00 PST. Thank you!

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 1]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
 in Kubernetes 1.26.
 {{< /note >}}
 
+## Changing the permissions of your's '/proc'
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:477]: The grammar doesn't look right to me.
I suggest:
```suggestion
## Managing access to the `/proc` filesystem {#proc-access}
```

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 2]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
...
+
+{{< feature-state for_k8s_version="v1.12" state="alpha" >}}
+
+By default, containers have multiple paths that are both
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:481]:
```suggestion
For runtimes that follow the OCI runtime specification, containers default to running in a mode where
there are multiple paths that are both
```

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 3]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
...
+[masked](https://github.com/opencontainers/runtime-spec/blob/f66aad47309/config-linux.md#masked-paths)
+and read only. The result of this is the container has these paths present
+inside the container's mount namespace, and they can function similarly to if
+the container was an isolated host, but the container process cannot write to
+them. A list of the paths are found in the [Kubernetes source
+code](https://github.com/kubernetes/kubernetes/blob/964529b227/pkg/securitycontext/util.go#L193-L211):
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:487]: Avoid linking to the source code; instead, if we want to put a list in, pop it inside a new (small) page within https://kubernetes.io/docs/reference/node/

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 4]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
...
+- Read Only Paths:
+  - `/proc/bus`
+  - `/proc/fs`
+  - `/proc/irq`
+  - `/proc/sys`
+  - `/proc/sysrq-trigger`
+
+However, there are some cases where a user may wish their container did not
+mask these paths. The most common case is users trying to run containers within
+Kubernetes containers.
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:510]: (nit) Try
```suggestion
For some Pods, you might want to bypass that default masking of paths.
The most common context for wanting this is if you are trying to run containers within
a Kubernetes container (within a pod).
```

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 5]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
 in Kubernetes 1.26.
 {{< /note >}}
 
+## Changing the permissions of your's '/proc'
+
+{{< feature-state for_k8s_version="v1.12" state="alpha" >}}
+
+By default, containers have multiple paths that are both
+[masked](https://github.com/opencontainers/runtime-spec/blob/f66aad47309/config-linux.md#masked-paths)
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:482]: I wouldn't put that link here, although I **would** put it inside the further reading section.

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 6]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
...
+  - `/proc/asound`
+  - `/proc/acpi`
+  - `/proc/kcore`
+  - `/proc/keys`
+  - `/proc/latency_stats`
+  - `/proc/timer_list`
+  - `/proc/timer_stats`
+  - `/proc/sched_debug`
+  - `/proc/scsi`
+  - `/sys/firmware`
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:499]: We might want to call out this one especially, as it's outside `/proc`.

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 7]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
...
+...
+securityContext:
+  procMount: Unmasked
+```
+
+{{ < note > }}
+As of Kubernetes 1.30, this field requires the `spec.hostUsers` value in the pod
+spec to be `false`. In other words: a container that wishes to have an Unmasked
+`/proc` must also be in a [user
+namespace](/docs/concepts/workloads/pods/user-namespaces.md).
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:525]:
```suggestion
{{ < note > }}
In Kubernetes {{< skew currentVersion >}}, this field requires the `spec.hostUsers` value in the pod
spec to be `false`. In other words: a container that wishes to have an Unmasked
`/proc` or unmasked `/sys` must also be in a
[user namespace](/docs/concepts/workloads/pods/user-namespaces/).

Kubernetes v1.12 to v1.29 did not enforce that requirement.
{{< /note >}}
```

If I want to map `/sys/firmware` into a (privileged) Pod, and I'm using Kubernetes v1.30, how do I do that?

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 8]
```diff
@@ -520,3 +570,4 @@ kubectl delete pod security-context-demo-4
 * For more information about security mechanisms in Linux, see
   [Overview of Linux Kernel Security Features](https://www.linux.com/learn/overview-linux-kernel-security-features)
   (Note: Some information is out of date)
+* [User Namespaces](/docs/concepts/workloads/pods/user-namespaces.md).
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:573]:
```suggestion
* Read about [user namespaces](/docs/concepts/workloads/pods/user-namespaces/) for Linux pods.
```

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 9]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
...
+mask these paths. The most common case is users trying to run containers within
+Kubernetes containers.
+
+The `securityContext` field `procMount` allows a user to request a container's `/proc`
+be `Unmasked`, or be mounted as read-write by the container process.
+
+```yaml
+...
+securityContext:
+  procMount: Unmasked
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:518]: We should give the reader more context about what to change; this looks pretty terse.

It's OK for alpha but would need to change before graduation to beta.

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 10]
```diff
@@ -474,6 +474,56 @@ can lose track of volume labels after restart. This deficiency has been fixed
 in Kubernetes 1.26.
 {{< /note >}}
 
+## Changing the permissions of your's '/proc'
+
+{{< feature-state for_k8s_version="v1.12" state="alpha" >}}
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:479]: Even better:
```suggestion
{{< feature-state feature_gate_name="ProcMountType" >}}
```

**Review by sftim — COMMENTED** [review]
Thanks; it'll need a few changes before we can merge this.

**Comment by sftim** [comment]
Please also fix the merge conflict (rebase against _dev-1.30_)

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 10 above)
Comment (Vyom-Yadav) [reply content/en/docs/tasks/configure-pod-container/security-context.md:479]: @haircommander I don't want to add unnecessary noise, but I thought [KEP-4265](https://github.com/kubernetes/enhancements/issues/4265) is targeting `beta`. Am I missing something over here? Are these the missing docs for `alpha`? Where are the docs for `beta`?

**Comment by Vyom-Yadav** [comment]
Hello @mrunalp :wave: please provide a technical review for this PR by March 26th 2024 18:00 PT to get this into the release. Thank you!

**Comment by Vyom-Yadav** [comment]
Hey @haircommander @mrunalp 👋🏼 

I'm reaching out from the Docs team. Just checking in as we approach Docs Freeze on `Tuesday March 26th 18:00 PDT`. This documentation appears to still be under review. To meet the Docs Freeze, this PR must have a technical review as well as `lgtm` and `approve` labels applied, without any unaddressed comments or concerns from SIG Docs. The status of this enhancement is marked as `at risk for docs freeze`. Thank you!

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 7 above)
Comment (haircommander) [reply content/en/docs/tasks/configure-pod-container/security-context.md:537]: if a pod is privileged then no paths are masked

**Review by mrunalp — APPROVED** [review]
/approve
Technically LGTM from sig-node.

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 3 above)
Comment (haircommander) [reply content/en/docs/tasks/configure-pod-container/security-context.md]: how do you feel about me dropping the link to the source code but not adding a reference? having it in 3 places seems more likely to skew eventually

**Comment by drewhagen** [comment]
Just a heads up that Doc Freeze is approaching fast tomorrow March 26th at 18:00 PDT and starting this cycle, we need to file [an Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) after this time.
This one looks pretty close and we'd like to help you get it over the finish line!

@kubernetes/sig-docs-en-owners Does this one look good for an `approve` from y'all? Thanks!

**Review by dipesh-rawat — COMMENTED** [review]
@haircommander The PR is currently lacking the doc change to transition the feature gate from Alpha to Beta. The necessary change should be made in the feature gate description file - `proc-mount-type.md` ([here](https://github.com/kubernetes/website/blob/dev-1.30/content/en/docs/reference/command-line-tools-reference/feature-gates/proc-mount-type.md)).

Please refer to the documentation [here](https://kubernetes.io/docs/contribute/new-content/new-features/#ready-for-review-feature-gates) for further information.

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 10 above)
Comment (Vyom-Yadav) [reply content/en/docs/tasks/configure-pod-container/security-context.md]: Oh, I see. The alpha feature was merged quite some time ago, but there weren't any docs for the feature. This PR just simply adds docs for that feature gate.

**Comment by Vyom-Yadav** [comment]
> The PR is currently lacking the doc change to transition the feature gate from Alpha to Beta.

@dipesh-rawat I don't think PR is targeting beta. [KEP-4265](https://github.com/kubernetes/enhancements/issues/4265) was updated to `stage/alpha`. So the feature gate will not be bumped. Rest @haircommander can confirm.

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 11]
```diff
@@ -477,6 +477,65 @@ runtime  recursively changes the SELinux label for all inodes (files and directo
 in the volume.
 The more files and directories in the volume, the longer that relabelling takes.
 
+{{< note >}}
+<!-- remove after Kubernetes v1.30 is released -->
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:481]: Is this right?

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 12]
```diff
@@ -477,6 +477,65 @@ runtime  recursively changes the SELinux label for all inodes (files and directo
...
+If you are running Kubernetes v1.25, refer to the v1.25 version of this task page:
+[Configure a Security Context for a Pod or Container](https://v1-25.docs.kubernetes.io/docs/tasks/configure-pod-container/security-context/) (v1.25).  
+There is an important note in that documentation about a situation where the kubelet
+can lose track of volume labels after restart. This deficiency has been fixed
+in Kubernetes 1.26.
+{{< /note >}}
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:487]:
```suggestion
```

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 7 above)
Comment (sftim) [reply content/en/docs/tasks/configure-pod-container/security-context.md:537]: OK, how about
```suggestion
{{ < note > }}
Setting `procMount` to Unmasked requires the `spec.hostUsers` value in the pod
spec to be `false`. In other words: a container that wishes to have an Unmasked
`/proc` or unmasked `/sys` must also be in a
[user namespace](/docs/concepts/workloads/pods/user-namespaces/).

Kubernetes v1.12 to v1.29 did not enforce that requirement.
{{< /note >}}
```
?

**Comment by sftim** [comment]
Also see https://github.com/kubernetes/website/pull/45152#discussion_r1539276457

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 3 above)
Comment (sftim) [reply content/en/docs/tasks/configure-pod-container/security-context.md]: That's fine.

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/tasks/configure-pod-container/security-context.md:509]: What I mean is: people might think that  `procMount` only covers `/proc`.

Comment (haircommander) [reply content/en/docs/tasks/configure-pod-container/security-context.md:509]: I left a note below does that suffice?

**Comment by haircommander** [comment]
> > The PR is currently lacking the doc change to transition the feature gate from Alpha to Beta.
> 
> @dipesh-rawat I don't think PR is targeting beta. [KEP-4265](https://github.com/kubernetes/enhancements/issues/4265) was updated to `stage/alpha`. So the feature gate will not be bumped. Rest @haircommander can confirm.

correct! we're not bumping this to beta in this release

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 11 above)
Comment (haircommander) [reply content/en/docs/tasks/configure-pod-container/security-context.md:481]: ah this was pulled in from a merge conflict...

File: content/en/docs/tasks/configure-pod-container/security-context.md (hunk 7 above)
Comment (drewhagen) [reply content/en/docs/tasks/configure-pod-container/security-context.md]: Looking at the diff, it looks like this change was made; this can be resolved.

**Comment by drewhagen** [comment]
Looks like feedback was addressed. @kubernetes/sig-docs-en-owners Is this ready to approve?

**Comment by reylejano** [comment]
Looks like feed back has been addressed and there is a technical lgtm from SIG Node
/approve

File: content/en/docs/tasks/configure-pod-container/security-context.md [hunk 13]
```diff
@@ -523,3 +575,7 @@ kubectl delete pod security-context-demo-4
 * For more information about security mechanisms in Linux, see
   [Overview of Linux Kernel Security Features](https://www.linux.com/learn/overview-linux-kernel-security-features)
   (Note: Some information is out of date)
+* Read about [User Namespaces](/docs/concepts/workloads/pods/user-namespaces.md)
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/security-context.md:578]: Ideally we'd have fixed this broken link before a merge. I'll file an issue.

Comment (sftim) [reply content/en/docs/tasks/configure-pod-container/security-context.md:578]: Actually, I'll just send in a PR.

**Pull request merged**